File: src/em_based/test/test_em.c
#include "em_gmm.h"

void test_list_fullcov(double *x, double *mus, double *covs, int *pM, int *pp) {
  int M = *pM, p = *pp;
  int m, i, j;
  params_link pl, params_fc;
  double loglik;

  params_fc = init_list_fullcov(mus, covs, M, p);

  pl = params_fc;
  for (m = 0; m < M; m++) {
    printf("%d: Mu = ", m);
    for (j = 0; j < p; j++)
      printf("%.3f ", pl->mu[j]);
    printf("\n");

    printf("%d : Cov\n", m);
    for (i = 0; i < p; i++) {
      for (j = 0; j < p; j++)
        printf("%.3f ", pl->cov[i * p + j]);
      printf("\n");
    }
    pl = pl->next;
  }

  pl = params_fc;
  for (m = 0; m < M; m++) {
    loglik = mvnorm_fullcov(x, pl, p);
    printf("loglik = %.5f\n", loglik);
    pl = pl->next;
  }

  free_list_fullcov(params_fc);
}

void test_list_diagcov(double *x, double *mus, double *sigs, int *pM, int *pp) {
  int M = *pM, p = *pp;
  int m, i, j;
  params_link pl, params_dc;
  double loglik;

  params_dc = init_list_diagcov(mus, sigs, M, p);

  pl = params_dc;
  for (m = 0; m < M; m++) {
    printf("%d: mus = ", m);
    for (j = 0; j < p; j++)
      printf("%.3f ", pl->mu[j]);
    printf("\n");

    printf("%d : sigs = ", m);
    for (j = 0; j < p; j++)
      printf("%.3f ", pl->sig[j]);
    printf("\n");

    pl = pl->next;
  }

  pl = params_dc;
  for (m = 0; m < M; m++) {
    loglik = mvnorm_diagcov(x, pl, p);
    printf("loglik = %.5f\n", loglik);
    pl = pl->next;
  }

  free_list_diagcov(params_dc);
}

void test_estep_fullcov(double *data, double *weights, double *props, double *mus, double *covs, int *pn, int *pM, int *pp) {
  int n = *pn, M = *pM, p = *pp;
  int m, i, j;
  double loglik;
  params_link params_fc;

  params_fc = init_list_fullcov(mus, covs, M, p);

  estep(params_fc, mvnorm_fullcov, data, weights, props, n, p, M);

  free_list_fullcov(params_fc);
}

void test_estep_diagcov(double *data, double *weights, double *props, double *mus, double *sigs, int *pn, int *pM, int *pp) {
  int n = *pn, M = *pM, p = *pp;
  int m, i, j;
  double loglik;
  params_link params_dc;

  params_dc = init_list_diagcov(mus, sigs, M, p);

  estep(params_dc, mvnorm_diagcov, data, weights, props, n, p, M);

  free_list_diagcov(params_dc);
}


void test_mstep_fullcov(double *data, double *weights, double *props, double *mus, double *covs, int *pn, int *pM, int *pp) {
  int n = *pn, M = *pM, p = *pp;
  int m, i, j;
  double loglik, sum;

  params_link pl, params_fc;

  params_fc = init_list_fullcov(mus, covs, M, p);

  estep(params_fc, mvnorm_fullcov, data, weights, props, n, p, M);

  pl = params_fc;
  for (m = 0; m < M; m++) {
    sum = mstep_fullcov(pl, data, &weights[m*n], n, p);
    props[m] = sum / n;
    pl = pl->next;
  }

  free_list_fullcov(params_fc);
}

void test_mstep_diagcov(double *data, double *weights, double *props, double *mus, double *covs, int *pn, int *pM, int *pp) {
  int n = *pn, M = *pM, p = *pp;
  int m, i, j;
  double loglik, sum;

  params_link pl, params_dc;

  params_dc = init_list_diagcov(mus, covs, M, p);

  estep(params_dc, mvnorm_diagcov, data, weights, props, n, p, M);

  pl = params_dc;
  for (m = 0; m < M; m++) {
    sum = mstep_diagcov(pl, data, &weights[m*n], n, p);
    props[m] = sum / n;
    pl = pl->next;
  }

  free_list_diagcov(params_dc);
}

File: src/em_based/test/test_em.R
## create share object using the following code in your bash
##  R CMD SHLIB -c test_em.c em_gmm.c
dyn.load("test_em.so")

## Test for EM

test_list_fullcov <- function(x, mus, covs) {
    M <- nrow(mus)
    p <- ncol(covs)

    ret <- .C("test_list_fullcov",
              x = as.double(x),
              mus = as.double(t(mus)),
              covs = as.double(covs),
              M = as.integer(M),
              p = as.integer(p))
}

M <- 3
Q <- MixSim::MixSim(0.05, K=M, p=2, sph=F, hom=F)
x <- matrix(c(0.3, 0.3), 1, 2)

test_list_fullcov(x, Q$Mu, Q$S)
for (m in 1:M) print(mclust::dmvnorm(x, Q$Mu[m,], Q$S[,,m], log=T))

test_list_diagcov <- function(x, mus, sigs) {
    M <- nrow(mus)
    p <- ncol(mus)

    ret <- .C("test_list_diagcov",
              x = as.double(x),
              mus = as.double(t(mus)),
              covs = as.double(t(sigs)),
              M = as.integer(M),
              p = as.integer(p))
}

M <- 3
Q <- MixSim::MixSim(0.05, K=M, p=2, sph=T, hom=F)
x <- matrix(c(0.3, 0.3), 1, 2)

s <- Q$Mu
for (m in 1:M) s[m, ] <- diag(Q$S[,,m])

test_list_diagcov(x, Q$Mu, s)
for (m in 1:M) print(mclust::dmvnorm(x, Q$Mu[m,], Q$S[,,m], log=T))

test_estep_fullcov <- function(X, props, mus, covs) {
    n <- nrow(X)
    p <- ncol(X)

    M <- nrow(mus)

    weights <- matrix(0, n, M)

    ret <- .C("test_estep_fullcov",
              data = as.double(t(X)),
              weights = as.double(weights),
              props = as.double(props),
              mus = as.double(t(mus)),
              covs = as.double(covs),
              n = as.integer(n),
              M = as.integer(M),
              p = as.integer(p))
    return(matrix(ret$weights, n, M))
}

## Estep test
n <- 3000
M <- 3
p <- 2
Q <- MixSim::MixSim(0.05, K=M, p=p, sph=F, hom=F)
A <- MixSim::simdataset(n, Pi=Q$Pi, Mu=Q$Mu, S=Q$S)
X <- A$X

Z <- matrix(0, n, M)
mus <- X[sample(1:n, M), ]
covs <- array(cov(X), dim=c(p, p, M))
props <- rep(1/M, M)

for (m in 1:M) Z[,m] <- props[m] * mclust::dmvnorm(X, mus[m, ], covs[,,m])
Z <- Z / rowSums(Z)

ret <- test_estep_fullcov(X, props, mus, covs)

norm(Z - ret, "2")

test_estep_diagcov <- function(X, props, mus, sigs) {
    n <- nrow(X)
    p <- ncol(X)

    M <- nrow(mus)

    weights <- matrix(0, n, M)

    ret <- .C("test_estep_diagcov",
              data = as.double(t(X)),
              weights = as.double(weights),
              props = as.double(props),
              mus = as.double(t(mus)),
              covs = as.double(t(sigs)),
              n = as.integer(n),
              M = as.integer(M),
              p = as.integer(p))
    return(matrix(ret$weights, n, M))
}

## Estep test
n <- 3000
M <- 3
p <- 2
Q <- MixSim::MixSim(0.05, K=M, p=p, sph=T, hom=F)
A <- MixSim::simdataset(n, Pi=Q$Pi, Mu=Q$Mu, S=Q$S)
X <- A$X

Z <- matrix(0, n, M)
mus <- X[sample(1:n, M), ]
covs <- Q$S
props <- rep(1/M, M)

for (m in 1:M) Z[,m] <- props[m] * mclust::dmvnorm(X, mus[m, ], covs[,,m])
Z <- Z / rowSums(Z)

s <- Q$Mu
for (m in 1:M) s[m, ] <- diag(covs[,,m])

ret <- test_estep_diagcov(X, props, mus, s)

norm(Z - ret, "2")

test_mstep_fullcov <- function(X, props, mus, covs) {
    n <- nrow(X)
    p <- ncol(X)
    M <- nrow(mus)

    weights <- matrix(0, n, M)

    ret <- .C("test_mstep_fullcov",
              data = as.double(t(X)),
              weights = as.double(weights),
              props = as.double(props),
              mus = as.double(t(mus)),
              covs = as.double(covs),
              n = as.integer(n),
              M = as.integer(M),
              p = as.integer(p))

    return(list(props=ret$props,
                mus=matrix(ret$mus, M, p, byrow=T),
                covs=array(ret$covs, dim=c(p, p, M))))

}

## Mstep test
n <- 3000
M <- 3
p <- 2
Q <- MixSim::MixSim(0.05, K=M, p=p, sph=F, hom=F)
A <- MixSim::simdataset(n, Pi=Q$Pi, Mu=Q$Mu, S=Q$S)
X <- A$X

Z <- matrix(0, n, M)
mus <- X[sample(1:n, M), ]
covs <- array(cov(X), dim=c(p, p, M))
props <- rep(1/M, M)

for (m in 1:M) Z[,m] <- props[m] * mclust::dmvnorm(X, mus[m, ], covs[,,m])
Z <- Z / rowSums(Z)

mus_hat <- t(Z) %*% X / colSums(Z)
covs_hat <- covs
for (m in 1:M) {
    tmp <- t(X) - mus_hat[m, ]
    covs_hat[,,m] <- tmp %*% (Z[, m] * t(tmp)) / sum(Z[, m])
}
props_hat <- colSums(Z) / n

ret <- test_mstep_fullcov(X, props, mus, covs)

norm(ret$mus - mus_hat, "2")
norm(ret$covs - covs_hat, "2")
norm(ret$props - props_hat, "2")

test_mstep_diagcov <- function(X, props, mus, sigs) {
    n <- nrow(X)
    p <- ncol(X)
    M <- nrow(mus)

    weights <- matrix(0, n, M)

    ret <- .C("test_mstep_diagcov",
              data = as.double(t(X)),
              weights = as.double(weights),
              props = as.double(props),
              mus = as.double(t(mus)),
              sigs = as.double(t(sigs)),
              n = as.integer(n),
              M = as.integer(M),
              p = as.integer(p))

    return(list(props=ret$props,
                mus=matrix(ret$mus, M, p, byrow=T),
                sigs=matrix(ret$sigs, M, p, byrow=T)
                )
           )
}

## Mstep test
n <- 3000
M <- 3
p <- 2
Q <- MixSim::MixSim(0.05, K=M, p=p, sph=T, hom=F)
A <- MixSim::simdataset(n, Pi=Q$Pi, Mu=Q$Mu, S=Q$S)
X <- A$X

Z <- matrix(0, n, M)
mus <- X[sample(1:n, M), ]
covs <- array(diag(p), dim=c(p, p, M))
props <- rep(1/M, M)

for (m in 1:M) Z[,m] <- props[m] * mclust::dmvnorm(X, mus[m, ], covs[,,m])
Z <- Z / rowSums(Z)

mus_hat <- t(Z) %*% X / colSums(Z)
covs_hat <- covs
for (m in 1:M) {
    tmp <- t(X) - mus_hat[m, ]
    covs_hat[,,m] <- tmp %*% (Z[, m] * t(tmp)) / sum(Z[, m])
}
props_hat <- colSums(Z) / n

sigs_hat <- mus_hat
for (m in 1:M) sigs_hat[m, ] <- diag(covs_hat[,,m])

s <- mus
for (m in 1:M) s[m, ] <- diag(covs[,,m])

ret <- test_mstep_diagcov(X, props, mus, s)

norm(ret$mus - mus_hat, "2")
norm(ret$sigs - sigs_hat, "2")
norm(ret$props - props_hat, "2")

em_gmm_fullcov <- function(X, M, silent=0, itmax=500, th=1e-4) {
    if (is.vector(X))
        X <- matrix(X, length(X), 1)
    n <- nrow(X)
    p <- ncol(X)

    dm <- p*(p+3)/2 ## the number of parameters

    mus <- X[sample(1:n, M),,drop=F]
    props <- rep(1, M) / M

    covs <- array(diag(p), dim=c(p, p, M))

    weights <- matrix(0, n, M)
    logliks <- rep(0, itmax)
    ret <- .C("em_gmm",
              data = as.double(t(X)),
              weights = as.double(weights),
              mus = as.double(t(mus)),
              covs = as.double(covs),
              props = as.double(props),
              cov_type = as.integer(0),
              n = as.integer(n),
              p = as.integer(p),
              M = as.integer(M),
              th = as.double(th),
              countf = integer(1),
              logliks = as.double(logliks),
              itmax = as.integer(itmax),
              silent = as.integer(silent))

    countf <- ret$countf

    logliks <- ret$logliks[1:countf]
    loglik <- tail(logliks, 1)

    weights <- matrix(ret$weights, n, M)

    mus <- matrix(ret$mus, M, p, byrow=T)
    covs <- array(ret$covs, dim=c(p, p, M))
    props <- ret$props

    list(weights=weights, props=props, mus=mus, covs=covs,
         logliks=logliks, loglik=loglik, countf=countf, cov_type = ret$cov_type)
}

## EM fullcov
n <- 3000
M <- 3
p <- 2
Q <- MixSim::MixSim(0.05, K=M, p=p, sph=T, hom=F)
A <- MixSim::simdataset(n, Pi=Q$Pi, Mu=Q$Mu, S=Q$S)
X <- A$X

ret <- em_gmm_fullcov(X, M, silent=0, itmax=1000, th=1e-8)

props <- ret$props
mus <- ret$mus
covs <- ret$covs

Z <- matrix(0, n, M)
for (m in 1:M) Z[,m] <- props[m] * mclust::dmvnorm(X, mus[m, ], covs[,,m])

sum(log(rowSums(Z))) - ret$loglik

em_gmm_diagcov <- function(X, M, silent=0, itmax=500, th=1e-4) {
    if (is.vector(X))
        X <- matrix(X, length(X), 1)
    n <- nrow(X)
    p <- ncol(X)

    mus <- X[sample(1:n, M),,drop=F]
    props <- rep(1, M) / M

    sigs <- matrix(1, M, p)

    weights <- matrix(0, n, M)
    logliks <- rep(0, itmax)
    ret <- .C("em_gmm",
              data = as.double(t(X)),
              weights = as.double(weights),
              mus = as.double(t(mus)),
              sigs = as.double(t(sigs)),
              props = as.double(props),
              cov_type = as.integer(1),
              n = as.integer(n),
              p = as.integer(p),
              M = as.integer(M),
              th = as.double(th),
              countf = integer(1),
              logliks = as.double(logliks),
              itmax = as.integer(itmax),
              silent = as.integer(silent))

    countf <- ret$countf

    logliks <- ret$logliks[1:countf]
    loglik <- tail(logliks, 1)

    weights <- matrix(ret$weights, n, M)

    mus <- matrix(ret$mus, M, p, byrow=T)
    sigs <- matrix(ret$sigs, M, p, byrow=T)
    props <- ret$props

    list(weights=weights, props=props, mus=mus, sigs=sigs,
         logliks=logliks, loglik=loglik, countf=countf, cov_type=ret$cov_type)
}

## EM diagcov
n <- 3000
M <- 3
p <- 2
Q <- MixSim::MixSim(0.05, K=M, p=p, sph=T, hom=F)
A <- MixSim::simdataset(n, Pi=Q$Pi, Mu=Q$Mu, S=Q$S)
X <- A$X

ret <- em_gmm_diagcov(X, M, silent=0, itmax=1000, th=1e-8)

props <- ret$props
mus <- ret$mus
sigs <- ret$sigs

covs <- array(0, dim=c(p, p, M))
for (m in 1:M) covs[,,m] <- diag(sigs[m,])

Z <- matrix(0, n, M)
for (m in 1:M) Z[,m] <- props[m] * mclust::dmvnorm(X, mus[m, ], covs[,,m])

sum(log(rowSums(Z))) - ret$loglik
